Split install page (#268)
Qiskit/documentation: docs/start/configure-qiskit-local.mdx

---
title: Configure Qiskit locally
description: Configure Qiskit on your local machine

---
# Configure Qiskit locally

After Qiskit is installed and running, there are some optional steps you can take to change the default Qiskit behavior.

## User configuration file

The main location for local configuration of Qiskit is the user configuration file. This is an .ini-format file that can be used to change Qiskit default settings.

Example:

 ```text
[default]
circuit_drawer = mpl
circuit_mpl_style = default
circuit_mpl_style_path = ~:~/.qiskit
state_drawer = hinton
transpile_optimization_level = 3
parallel = False
num_processes = 15
 ```

> By default this file is in `~/.qiskit/settings.conf` but the path can be overridden with the QISKIT_SETTINGS environment variable.

Reviewer: ```suggestion
By default, this file is in `~/.qiskit/settings.conf` but the path can be overridden with the QISKIT_SETTINGS environment variable.
```

## Available options

- `circuit_drawer`: Changes the default backend for the circuit drawer. It can be set to `latex`, `mpl`, `text`, or `latex_source`. When the output kwarg is not explicitly set, this drawer backend is used.
- `circuit_mpl_style`: The default style sheet used for the mpl output backend for the circuit drawer. Valid values are `default` or `bw`.
- `circuit_mpl_style_path`: The paths to have the circuit drawer use to look for JSON style sheets when using the mpl output mode.

> - `state_drawer`: This is used to change the default backend for the state visualization draw methods. Valid values are `repr`, `text`, `latex`, `latex_source`, `qsphere`, `hinton`, or `bloch`. When the output kwarg is not explicitly set on the :meth:`~qiskit.quantum_info.DensityMatrix.draw` method, the specified output method is used.

Reviewer: Are we supposed to be able to see the :meth: tags  and the prepended ~ throughout? Or is this leftover formatting from the migration?

Author: I guess?  I thought it looked weird, but I checked and you can see them in the original link I was sent too: https://github.com/Qiskit/qiskit/blob/main/docs/configuration.rst#id1

Reviewer: Hm. Looks like shards of .rst to me... I found [this page](https://stackoverflow.com/questions/34848253/whats-the-difference-in-behaviour-between-func-and-meth-roles-in-python-sph) that seems to suggest that the html output shouldn't show the tags (also for :func:)??

Reviewer 2: Oh yeah those :meth: things definitely shouldnt be there, it must have been an error in the originaly doc, instead it should be a link to that method in the api ref

Author: I tried to add links for all of them; hopefully they're going to the right spot.

- `transpile_optimization_level`: Change the default optimization level for :func:`~qiskit.compiler.transpile` and :func:`~qiskit.execute.execute`. Specify an integer 0-3.
- `parallel`: Whether Python multiprocessing is enabled for operations that support running in parallel.  For example, transpilation of multiple :class:`~qiskit.circuit.QuantumCircuit` objects. This setting can be overridden by the QISKIT_PARALLEL environment variable. Specify a boolean value.
- `num_processes`: The maximum number of parallel processes to launch for parallel operations if parallel execution is enabled. This setting can be overridden by the QISKIT_NUM_PROCS environment variable. Specify an integer greater than 0. 

<Admonition type="note">
    * Circuit drawer settings apply to :meth:`qiskit.circuit.QuantumCircuit.draw` and :func:`qiskit.visualization.circuit_drawer`.
    * State visualization draw methods are :meth:`qiskit.quantum_info.Statevector.draw` and :meth:`qiskit.quantum_info.DensityMatrix.draw`
</Admonition>

> ## Environment variables

Reviewer: I wonder if environment variables should be in code font? Do you know if there's a standard @beckykd @javabster ?

Reviewer 2: I think its fine as is :)

Author: I looked at IBM Style and there's no guidance, but then I looked at some other IBM files that describe environment variables and none of them have special formatting.

> Setting these environment variables alters the default behavior of Qiskit:

Reviewer: ```suggestion
Set these environment variables to alter the default behavior of Qiskit:
```

> - QISKIT_PARALLEL: Enables Python multiprocessing to parallelize certain operations, for example, transpilation over multiple circuits in Qiskit. Specify a boolean value.

Reviewer: ```suggestion
- QISKIT_PARALLEL: Enables Python multiprocessing to parallelize certain operations; for example, transpilation over multiple circuits in Qiskit. Specify a boolean value.
```

- QISKIT_NUM_PROCS: The maximum number of parallel processes to launch for parallel operations if parallel execution is enabled. Specify an integer greater than zero.

> - RAYON_NUM_THREADS: The number of threads to run multithreaded operations in Qiskit. By default multithreaded code launches a thread for each logical CPU. To adjust the number of threads Qiskit uses, set this to an integer value. For example, setting RAYON_NUM_THREADS=4 Launches four threads for multithreaded functions. 

Reviewer: ```suggestion
- RAYON_NUM_THREADS: The number of threads to run multithreaded operations in Qiskit. By default multithreaded code launches a thread for each logical CPU. To adjust the number of threads Qiskit uses, set this to an integer value. For example, setting RAYON_NUM_THREADS=4 launches four threads for multithreaded functions. 
```

> - QISKIT_FORCE_THREADS: Specifies that multithreaded code should always execute in multiple threads. By default if you're running multithreaded code in a section of Qiskit that is already running in parallel processes, Qiskit does not launch multiple threads but instead executes that function serially. This is done to avoid potentially overloading limited CPU resources. However, if you want to force the use of multiple threads even when in a multiprocess context, set QISKIT_FORCE_THREADS=TRUE.

Reviewer: ```suggestion
- QISKIT_FORCE_THREADS: Specifies that multithreaded code should always execute in multiple threads. By default, if you're running multithreaded code in a section of Qiskit that is already running in parallel processes, Qiskit does not launch multiple threads but instead executes that function serially. This is done to avoid potentially overloading limited CPU resources. However, if you want to force the use of multiple threads even when in a multiprocess context, set QISKIT_FORCE_THREADS=TRUE.
```
